modules/netstat/netstat_linux.go

```diff
@@ -9,15 +9,152 @@ package netstat /* import "mig.ninja/mig/modules/netstat" */
 import (
 	"bufio"
 	"fmt"
+	"io/ioutil"
 	"net"
 	"os"
+	"path"
 	"regexp"
 	"strconv"
 	"strings"
 )
 
-// HasSeenMac on linux looks for a matching mac address in /proc/net/arp
-// and returns its MAC and IP address if found
+// Represents lines obtained from the /proc file system related to network
+// activity
+type procNetLine struct {
+	line         string // The line from the proc file
+	nsIdentifier string // A namespace identifier associated with the process
+}
+
+// Return process candidates organized by active network namespace, this will
+// include the host namespace in addition to other namespaces
+func procCandidates() (map[string][]string, error) {
+	var err error
+	ret := make(map[string][]string)
+
+	dirents, err := ioutil.ReadDir("/proc")
+	if err != nil {
+		return ret, err
+	}
+	for _, x := range dirents {
+		_, err = strconv.Atoi(x.Name())
+		if err != nil {
+			continue
+		}
+		nspath := path.Join("/proc", x.Name(), "ns", "net")
+		nsname, err := os.Readlink(nspath)
+		if err != nil {
+			// We don't treat this condition as fatal
+			continue
+		}
+		if _, ok := ret[nsname]; !ok {
+			ret[nsname] = make([]string, 0)
+		}
+		ret[nsname] = append(ret[nsname], path.Join("/proc", x.Name(), "net"))
+	}
+
+	return ret, nil
+}
+
+// Read lines from a network proc file; used by procNetNS()
+func procNetNSEntries(fname string) (ret []string, err error) {
+	fd, err := os.Open(fname)
+	if err != nil {
+		return ret, err
+	}
+	defer fd.Close()
+	scanner := bufio.NewScanner(fd)
+	scanner.Scan() // Skip the header
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return ret, err
+		}
+		ret = append(ret, scanner.Text())
+	}
+	return ret, nil
+}
+
+// Given a network proc file (e.g., tcp would represent instances of
+// <pid>/net/tcp visible by the agent, return a list of proc network entries
+// for each active network namespace, including the host.
+//
+// This function works by building a cache of all current network namespaces
+// on the host based on /proc. From this list, we try to collect data from the
+// indicated file for at most one process per namespace. Since <pid>/net/*
+// represents network activity for the namespace the process is in, and not
+// the process itself we only need to be successful with one PID in each
+// namespace.
+func procNetNS(fname string) (ret []procNetLine, err error) {
```

Comment:
What about using a channel that `procNetNS` writes `procNetLine`s into, and the caller functions can read from the scan the files? That would circumvent the need for a slice of lines.